### DISCUSSION TITLE
Access pattern in PVMT (config and object)

### DISCUSSION THREAD

**Discussion by ewuerger** [dbinfrago/py-capellambse#255] [General]
With the PVMT rework the access will be granted on the model itself (configuration) and on any `GenericElement` (object) via `.pvmt`. Now from the previous implementation access to domains, groups and property-values on object level implemented dict-like behavior, e.g.:
```
group = obj.pvmt["Computer.Physical Cables"]
pv = obj.pvmt["Computer.Physical Cables.In Stock"]
``` 

Notice the `.` (dot) notation which controls what we end up with (either a group or a property-value). To minimize implementation and maintenance effort there should be only one, pythonic way to access wanted PVMT elements on config and object level. We want to agree on one way with a poll later.

Current poll options:

1. `obj.pvmt["domain.group.property_value"]`
2. `obj.pvmt["domain"]["group"]["property_value"]`
3. `obj.pvmt.domains.by_name("domain").groups.by_name("group").properties.by_name("property_value")`
4. `obj.pvmt.domain.group.property_value`
5. `obj.pvmt.domains["my-domain"].groups["my-group"]`

**_If no poll option satisfies your needs or you have a better option, please share it in a comment below. Your ideas will be added as a poll option._**

**Comment by ewuerger** [comment]
We agreed on the following pattern:
`obj.pvmt["domain"]["group"]["property_value"]`

and agreed on 4. (pandas inspired) as being nice-to-have. The latter can only work when the name is feasible as an attribute. We also discussed a feature where one can find any PVMT object (Domain, Group of ProperyValue) by name.

**Reply by Wuestengecko** [reply]
> and agreed on 4. (pandas inspired) as being nice-to-have

Did we? :thinking: We agreed that option 2 is the winner, but I don't recall an actual discussion about whether or not we want 4 _additionally_.

Considering the very strong constraint it sets on domain/group/prop names ("must be a valid Python identifier"), IMO it doesn't make sense to have it. Furthermore, it could lead to confusion between those "virtual" attributes and "real" attributes on the Python object, like "name" (which basically always exists) – is the user asking for the name of the group object while he's iterating over `model.search()`, or is he interested in a property value called "name" inside this group?

I'm definitely in for the unambiguous approach on this one, even if it takes 3 characters more (`[""]` vs. `.`).

**Reply by ewuerger** [reply]
Yes you are right.
